Could not pass event PotionSplashEvent to GriefPrevention

### What happened:
Error: Could not pass event PotionSplashEvent to GriefPrevention

### What was expected:
_What you expected to see_

### Steps to reproduce:
__This may include a build schematic, a video, or detailed instructions to help reconstruct the issue__
Use splash potion in claim owned by offline player?

### Server and GriefPrevention version:
___This can be found by running `/version` and `/version GriefPrevention` on your server.___
```> ver
[22:05:41 INFO]: Checking version, please wait...
[22:05:41 INFO]: This server is running Paper version git-Paper-302 (MC: 1.17.1) (Implementing API version 1.17.1-R0.1-SNAPSHOT) (Git: a978f41)
You are 9 version(s) behind
Previous version: git-Paper-299 (MC: 1.17.1)
> ver GriefPrevention
[22:05:44 INFO]: GriefPrevention version 16.17.1-30-ge6ce386-dirty
```

### Stack trace/error or server log
```
com.destroystokyo.paper.exception.ServerEventException: Could not pass event PotionSplashEvent to GriefPrevention v16.17.1-30-ge6ce386-dirty
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:645)
	at org.bukkit.craftbukkit.v1_17_R1.event.CraftEventFactory.callPotionSplashEvent(CraftEventFactory.java:793)
	at net.minecraft.world.entity.projectile.EntityPotion.a(EntityPotion.java:197)
	at net.minecraft.world.entity.projectile.EntityPotion.a(EntityPotion.java:115)
	at net.minecraft.world.entity.projectile.IProjectile.preOnHit(IProjectile.java:172)
	at net.minecraft.world.entity.projectile.EntityProjectile.tick(EntityProjectile.java:80)
	at net.minecraft.server.level.WorldServer.entityJoinedWorld(WorldServer.java:1114)
	at net.minecraft.world.level.World.a(World.java:986)
	at net.minecraft.server.level.WorldServer.lambda$tick$7(WorldServer.java:775)
	at net.minecraft.world.level.entity.EntityTickList.a(EntityTickList.java:43)
	at net.minecraft.server.level.WorldServer.doTick(WorldServer.java:756)
	at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1619)
	at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:480)
	at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1475)
	at net.minecraft.server.MinecraftServer.x(MinecraftServer.java:1274)
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:319)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.NullPointerException: Cannot invoke "org.bukkit.entity.Player.getUniqueId()" because "checked" is null
	at GriefPrevention.jar//me.ryanhamshire.GriefPrevention.events.ClaimPermissionCheckEvent.<init>(ClaimPermissionCheckEvent.java:39)
	at GriefPrevention.jar//me.ryanhamshire.GriefPrevention.Claim.checkPermission(Claim.java:458)
	at GriefPrevention.jar//me.ryanhamshire.GriefPrevention.EntityEventHandler.onPotionSplash(EntityEventHandler.java:1489)
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor347.execute(Unknown Source)
	at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69)
	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80)
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70)
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:626)
	at org.bukkit.craftbukkit.v1_17_R1.event.CraftEventFactory.callPotionSplashEvent(CraftEventFactory.java:793)
	at app//net.minecraft.world.entity.projectile.ThrownPotion.applySplash(ThrownPotion.java:197)
	at app//net.minecraft.world.entity.projectile.ThrownPotion.onHit(ThrownPotion.java:115)
	at app//net.minecraft.world.entity.projectile.Projectile.preOnHit(Projectile.java:172)
	at app//net.minecraft.world.entity.projectile.ThrowableProjectile.tick(ThrowableProjectile.java:80)
	at app//net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1114)
	at app//net.minecraft.world.level.Level.guardEntityTick(Level.java:986)
	at app//net.minecraft.server.level.ServerLevel.lambda$tick$7(ServerLevel.java:775)
	at app//net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:43)
	at app//net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:756)
	at app//net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1619)
	at app//net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:480)
	at app//net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1475)
	at app//net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1274)
	... 2 more

```
Or, paste the server.log at gist.github.com and paste the link here: 

### GriefPrevention config.yml

```yml

Paste the config.yml right here inside these backticks. Please do not use a paste service for this.

```

### Plugin list (if applicable):



### Debugging steps attempted
<!--These will be checkboxes you can click after creating the issue)-->
- [ ] (If applicable) I attempted running the server without GriefPrevention installed
  - [ ] (If applicable) The problem does _not_ occur if GriefPrevention is removed from the server.
- [ ] I attempted running only GriefPrevention on the server
  - [ ] Issue still occurs even if GriefPrevention is the only plugin running
- [ ] I attempted testing for the issue on a new server
  - [ ] Issue still occurs even after testing on a new server
- [ ] I attempted testing for the issue with the vanilla client (no mods)
  - [ ] Issue still occurs even after testing with the vanilla Minecraft client

--- a/src/main/java/me/ryanhamshire/GriefPrevention/EntityEventHandler.java
+++ b/src/main/java/me/ryanhamshire/GriefPrevention/EntityEventHandler.java
@@ -309,12 +309,17 @@ public class EntityEventHandler implements Listener
         }
 
         // Allow change if projectile was shot by a dispenser in the same claim.
-        if (shooter instanceof BlockProjectileSource &&
-                GriefPrevention.instance.dataStore.getClaimAt(((BlockProjectileSource) shooter).getBlock().getLocation(), false, claim) == claim)
+        if (isBlockSourceInClaim(shooter, claim))
             return;
 
         // Prevent change in all other cases.
         event.setCancelled(true);
+    }
+
+    private boolean isBlockSourceInClaim(ProjectileSource projectileSource, Claim claim)
+    {
+        return projectileSource instanceof BlockProjectileSource &&
+                GriefPrevention.instance.dataStore.getClaimAt(((BlockProjectileSource) projectileSource).getBlock().getLocation(), false, claim) == claim;
     }
 
     //Used by "sand cannon" fix to ignore fallingblocks that fell through End Portals
@@ -1461,37 +1466,62 @@ public class EntityEventHandler implements Listener
     {
         ThrownPotion potion = event.getPotion();
 
-        //ignore potions not thrown by players
         ProjectileSource projectileSource = potion.getShooter();
+        // Ignore potions with no source.
         if (projectileSource == null) return;
         Player thrower = null;
         if ((projectileSource instanceof Player))
             thrower = (Player) projectileSource;
+        boolean messagedPlayer = false;
 
         Collection<PotionEffect> effects = potion.getEffects();
         for (PotionEffect effect : effects)
         {
             PotionEffectType effectType = effect.getType();
 
-            //restrict some potions on claimed animals (griefers could use this to kill or steal animals over fences) //RoboMWM: include villagers
-            if (effectType.getName().equals("JUMP") || effectType.getName().equals("POISON"))
+            // Restrict some potions on claimed villagers and animals.
+            // Griefers could use potions to kill entities or steal them over fences.
+            if (PotionEffectType.HARM.equals(effectType)
+                    || PotionEffectType.POISON.equals(effectType)
+                    || PotionEffectType.JUMP.equals(effectType)
+                    || PotionEffectType.WITHER.equals(effectType))
             {
                 Claim cachedClaim = null;
-                for (LivingEntity effected : event.getAffectedEntities())
+                for (LivingEntity affected : event.getAffectedEntities())
                 {
-                    if (effected.getType() == EntityType.VILLAGER || effected instanceof Animals)
+                    // Always impact the thrower.
+                    if (affected == thrower) continue;
+
+                    if (affected.getType() == EntityType.VILLAGER || affected instanceof Animals)
                     {
-                        Claim claim = this.dataStore.getClaimAt(effected.getLocation(), false, cachedClaim);
+                        Claim claim = this.dataStore.getClaimAt(affected.getLocation(), false, cachedClaim);
                         if (claim != null)
                         {
                             cachedClaim = claim;
-                            Supplier<String> override = () -> instance.dataStore.getMessage(Messages.NoDamageClaimedEntity, claim.getOwnerName());
-                            final Supplier<String> noContainersReason = claim.checkPermission(thrower, ClaimPermission.Inventory, event, override);
-                            if (thrower == null || noContainersReason != null)
+
+                            if (thrower == null)
                             {
-                                event.setIntensity(effected, 0);
-                                GriefPrevention.sendMessage(thrower, TextMode.Err, noContainersReason.get());
-                                return;
+                                // Non-player source: Witches, dispensers, etc.
+                                if (!isBlockSourceInClaim(projectileSource, claim))
+                                {
+                                    // If the source is not a block in the same claim as the affected entity, disallow.
+                                    event.setIntensity(affected, 0);
+                                }
+                            }
+                            else
+                            {
+                                // Source is a player. Determine if they have permission to access entities in the claim.
+                                Supplier<String> override = () -> instance.dataStore.getMessage(Messages.NoDamageClaimedEntity, claim.getOwnerName());
+                                final Supplier<String> noContainersReason = claim.checkPermission(thrower, ClaimPermission.Inventory, event, override);
+                                if (noContainersReason != null)
+                                {
+                                    event.setIntensity(affected, 0);
+                                    if (!messagedPlayer)
+                                    {
+                                        GriefPrevention.sendMessage(thrower, TextMode.Err, noContainersReason.get());
+                                        messagedPlayer = true;
+                                    }
+                                }
                             }
                         }
                     }
@@ -1528,7 +1558,11 @@ public class EntityEventHandler implements Listener
                         if (!pvpEvent.isCancelled())
                         {
                             event.setIntensity(effected, 0);
-                            GriefPrevention.sendMessage(thrower, TextMode.Err, Messages.CantFightWhileImmune);
+                            if (!messagedPlayer)
+                            {
+                                GriefPrevention.sendMessage(thrower, TextMode.Err, Messages.CantFightWhileImmune);
+                                messagedPlayer = true;
+                            }
                         }
                         continue;
                     }
@@ -1542,7 +1576,11 @@ public class EntityEventHandler implements Listener
                         if (!pvpEvent.isCancelled())
                         {
                             event.setIntensity(effected, 0);
-                            GriefPrevention.sendMessage(thrower, TextMode.Err, Messages.PlayerInPvPSafeZone);
+                            if (!messagedPlayer)
+                            {
+                                GriefPrevention.sendMessage(thrower, TextMode.Err, Messages.PlayerInPvPSafeZone);
+                                messagedPlayer = true;
+                            }
                         }
                     }
                 }

